File: shm/uhash_loader.c
/* standalone uhash loader -- jochang */

#include "uhash_loader.h"

int load_uhash(void) {
    int shmid, err;
    fprintf(stderr, "load_uhash: start\n");
    shmid = shmget(SHM_KEY, SHMSIZE,
#ifdef USE_HUGETLB
		   SHM_HUGETLB |
#endif
		   0600 | IPC_CREAT | IPC_EXCL);
    err = errno;
    fprintf(stderr, "after shmget: err: %d EEXIST: %d SHM_KEY: %d SHMSIZE: %luSHM_t: %lu SHMALIGNEDSIZE: %d\n", err, EEXIST, SHM_KEY, SHMSIZE, sizeof(SHM_t), SHMALIGNEDSIZE);
    if( err == EEXIST )
	shmid = shmget(SHM_KEY, SHMSIZE,
#ifdef USE_HUGETLB
		       SHM_HUGETLB |
#endif
		       0600 | IPC_CREAT);

    fprintf(stderr, "after shmget-2: shmid: %d\n", shmid);
#ifdef USE_HUGETLB
    fprintf(stderr, "after shmget-2: use hugetlb\n");
#endif

    if( shmid < 0 ){
	fprintf(stderr, "[ERROR] shmget\n");
  return 1;
    }
    SHM = (void *) shmat(shmid, NULL, 0);
    if( SHM == (void *)-1 ){
	fprintf(stderr, "[ERROR] shmat\n");
  return 1;
    }
    if( err  != EEXIST ) {
	SHM->number=SHM->loaded = 0;
	SHM->version = SHM_VERSION;
	SHM->size    = sizeof(SHM_t);
    }

    if(SHM->version != SHM_VERSION) {
      fprintf(stderr, "Error: SHM->version(%d) != SHM_VERSION(%d)\n", SHM->version, SHM_VERSION);
      fprintf(stderr, "Please use the source code version corresponding to SHM,\n"
	 "or use ipcrm(1) command to clean share memory.\n");
      return 1;
    }

// in case it's not assumed zero, this becomes a race...
    if( SHM->number == 0 && SHM->loaded == 0 ){
	SHM->loaded = 0;
	err = fill_uhash(0);
	SHM->today_is[0] = '\0';
	SHM->loaded = 1;
    }
    else{
	err = fill_uhash(1);
    }

    return err;
}

void checkhash(int h)
{
    int *p = &(SHM->hash_head[h]), ch, deep=0;
    while(*p != -1)
    {
       if(*p <-1 || *p >= MAX_USERS) {*p=-1; return;}
       ch = StringHash( SHM->userid[*p])%(1<<HASH_BITS);
       if(ch!=h)
       {
           printf("remove %d %d!=%d %d [%s] next:%d\n",
		    deep, h, ch, *p, SHM->userid[*p],
		    SHM->next_in_hash[*p]);
          *p = SHM->next_in_hash[*p]; //remove from link
          // *p=-1;  Ptt: cut it?
	  //return;
       }
       else
          p = &(SHM->next_in_hash[*p]);
       deep++;
    }
}
int fill_uhash(int onfly)
{
    int fd, usernumber;
    usernumber = 0;

    for (fd = 0; fd < (1 << HASH_BITS); fd++)
      if(!onfly)
    	  SHM->hash_head[fd] = -1;
      else
	  checkhash(fd);

    if ((fd = open(FN_PASSWD, O_RDWR)) > 0)
    {
	struct stat stbuf;
	caddr_t fimage, mimage;

	fstat(fd, &stbuf);
	fimage = mmap(NULL, stbuf.st_size, PROT_WRITE|PROT_READ, MAP_SHARED, fd, 0);
	if (fimage == (char *) -1)
	{
	    fprintf(stderr,"[ERROR] mmap");
	    return 1;
	}
	close(fd);
	fd = stbuf.st_size / sizeof(userec_t);
	if (fd > MAX_USERS)
	    fd = MAX_USERS;

	for (mimage = fimage; usernumber < fd; mimage += sizeof(userec_t))
	{
	    userec_add_to_uhash(usernumber, (userec_t *)mimage, onfly);
	    usernumber++;
	}
	munmap(fimage, stbuf.st_size);
    }
    else
    {
	fprintf(stderr, "[ERROR] open");
	return 1;
    }
    SHM->number = usernumber;

    printf("total %d names %s.\n", usernumber, onfly ? "checked":"loaded");
    return 0;
}
void userec_add_to_uhash(int n, userec_t *user, int onfly)
{
    int *p, h, l=0;

    // uhash use userid="" to denote free slot for new register
    // However, such entries will have the same hash key.
    // So we skip most of invalid userid to prevent lots of hash collision.
    if (!is_validuserid(user->userid)) {
	// dirty hack, preserve few slot for new register
	static int count = 0;
	count++;
	if (count > 1000)
	    return;
    }

    h = StringHash(user->userid)%(1<<HASH_BITS);

    p = &(SHM->hash_head[h]);
    if(!onfly || SHM->userid[n][0] != user->userid[0] ||
	       strncmp(SHM->userid[n], user->userid, IDLEN-1))
    {
       strcpy(SHM->userid[n], user->userid);
       SHM->money[n] = user->money;
#ifdef USE_COOLDOWN
       SHM->cooldowntime[n] = 0;
#endif
       if(onfly)
           printf("add %s\n", user->userid);
    }
    while (*p != -1)
    {
	if(onfly && *p==n )  // already in hash
	     return;
	l++;
	p = &(SHM->next_in_hash[*p]);
    }
    if(onfly)
       printf("add %d %d %d [%s] in hash\n", l, h, n, user->userid);
    SHM->next_in_hash[*p = n] = -1;

    return;
}

File: cache/uhash_loader.go
package cache

import (
	"io"
	"os"
	"reflect"
	"unsafe"

	"github.com/PichuChen/go-bbs/cmbbs"
	"github.com/PichuChen/go-bbs/cmsys"
	"github.com/PichuChen/go-bbs/ptttype"
	"github.com/PichuChen/go-bbs/types"
	log "github.com/sirupsen/logrus"
)

//LoadUHash
//Load user-hash into SHM.
func LoadUHash() (err error) {
	if Shm == nil {
		return ErrShmNotInit
	}

	// line: 58
	number := int32(0)
	Shm.ReadAt(
		unsafe.Offsetof(Shm.Number),
		unsafe.Sizeof(Shm.Number),
		unsafe.Pointer(&number),
	)

	loaded := int32(0)
	Shm.ReadAt(
		unsafe.Offsetof(Shm.Loaded),
		unsafe.Sizeof(Shm.Loaded),
		unsafe.Pointer(&loaded),
	)

	//XXX in case it's not assumed zero, this becomes a race...
	if number == 0 && loaded == 0 {
		// line: 60
		fillUHash(false)

		// line: 61
		zeroByte := '\x00'
		Shm.WriteAt(
			unsafe.Offsetof(Shm.TodayIs),
			unsafe.Sizeof(Shm.TodayIs[0]),
			unsafe.Pointer(&zeroByte),
		)

		// line: 62
		loaded = 1
		Shm.WriteAt(
			unsafe.Offsetof(Shm.Loaded),
			unsafe.Sizeof(Shm.Loaded),
			unsafe.Pointer(&loaded),
		)
	} else {
		// line: 65
		fillUHash(true)
	}

	return nil
}

var (
	uHashLoaderInvalidUserID = 0
)

func fillUHash(isOnfly bool) error {
	log.Infof("fillUHash: start: isOnfly: %v", isOnfly)
	initFillUHash(isOnfly)

	file, err := os.Open(ptttype.FN_PASSWD)
	if err != nil {
		return err
	}

	usernum := int32(0)

	uHashLoaderInvalidUserID = 0
	for ; ; usernum++ {
		userecRaw, eachErr := ptttype.NewUserecRawWithFile(file)
		if eachErr != nil {
			// io.EOF is reading correctly to the end the file.
			if eachErr == io.EOF {
				break
			}

			err = eachErr
			break
		}

		userecRawAddToUHash(usernum, userecRaw, isOnfly)
	}

	Shm.WriteAt(
		unsafe.Offsetof(Shm.Number),
		unsafe.Sizeof(Shm.Number),
		unsafe.Pointer(&usernum),
	)
	return nil
}

func userecRawAddToUHash(usernum int32, userecRaw *ptttype.UserecRaw, isOnfly bool) {
	// uhash use userid="" to denote free slot for new register
	// However, such entries will have the same hash key.
	// So we skip most of invalid userid to prevent lots of hash collision.
	if !cmbbs.IsValidUserID(&userecRaw.UserID) {
		// dirty hack, preserve few slot for new register
		uHashLoaderInvalidUserID++
		if uHashLoaderInvalidUserID > 1000 {
			return
		}
	}

	h := cmsys.StringHash(userecRaw.UserID[:]) % (1 << ptttype.HASH_BITS)

	shmUserID := [ptttype.IDLEN + 1]byte{}
	Shm.ReadAt(
		unsafe.Offsetof(Shm.Userid)+ptttype.USER_ID_SZ*uintptr(usernum),
		ptttype.USER_ID_SZ,
		unsafe.Pointer(&shmUserID),
	)

	offsetNextInHash := unsafe.Offsetof(Shm.NextInHash)

	if !isOnfly || !reflect.DeepEqual(userecRaw.UserID, shmUserID) {
		Shm.WriteAt(
			unsafe.Offsetof(Shm.Userid)+ptttype.USER_ID_SZ*uintptr(usernum),
			ptttype.USER_ID_SZ,
			unsafe.Pointer(&userecRaw.UserID),
		)

		Shm.WriteAt(
			unsafe.Offsetof(Shm.Money)+types.INT32_SZ*uintptr(usernum),
			types.INT32_SZ,
			unsafe.Pointer(&userecRaw.Money),
		)

		if ptttype.USE_COOLDOWN {
			zero := types.Time4(0)
			Shm.WriteAt(
				unsafe.Offsetof(Shm.CooldownTime)+types.TIME4_SZ*uintptr(usernum),
				types.TIME4_SZ,
				unsafe.Pointer(&zero),
			)
		}
		log.Debugf("UHashLoader.userecRawAddToUHash: add info: usernum: %v id: %v shmUserID: %v", usernum, string(userecRaw.UserID[:]), string(shmUserID[:]))
	}

	p := h
	val := int32(0)
	offsetHashHead := unsafe.Offsetof(Shm.HashHead)
	//offsetNextInHash := unsafe.Offsetof(Shm.NextInHash)
	isFirst := true

	Shm.ReadAt(
		offsetHashHead+types.INT32_SZ*uintptr(p),
		types.INT32_SZ,
		unsafe.Pointer(&val),
	)

	l := 0
	for val >= 0 && val < ptttype.MAX_USERS {
		if isOnfly && val == usernum { // already in hash
			return
		}

		l++
		// go to next
		// 1. setting p as val
		// 2. get val from next_in_hash[p]
		p = uint32(val)
		Shm.ReadAt(
			offsetNextInHash+types.INT32_SZ*uintptr(p),
			types.INT32_SZ,
			unsafe.Pointer(&val),
		)

		isFirst = false
	}

	// set next in hash as n
	offset := offsetHashHead
	if !isFirst {
		offset = offsetNextInHash
	}
	val = usernum
	Shm.WriteAt(
		offset+types.INT32_SZ*uintptr(p),
		types.INT32_SZ,
		unsafe.Pointer(&val),
	)

	log.Infof("UHashLoader.userecRawAddToUHash: added level: %v p: %v hash: %v usernum: %v [%v] val: %v in hash isHashHead: %v", l, p, h, usernum, string(userecRaw.UserID[:]), val, isFirst)

	// set next in hash as -1
	p = uint32(val)
	val = -1
	Shm.WriteAt(
		offsetNextInHash+types.INT32_SZ*uintptr(p),
		types.INT32_SZ,
		unsafe.Pointer(&val),
	)
	log.Debugf("UHashLoader.userecRawAddToUHash: added NextInHash: usernum: %v p: %v val: %v isFirst: %v", usernum, p, val, isFirst)
}

func initFillUHash(isOnfly bool) {
	if !isOnfly {
		toFillHashHead := [1 << ptttype.HASH_BITS]int32{}
		for idx := range toFillHashHead {
			toFillHashHead[idx] = -1
		}
		Shm.WriteAt(
			unsafe.Offsetof(Shm.HashHead),
			unsafe.Sizeof(Shm.HashHead),
			unsafe.Pointer(&toFillHashHead),
		)
	} else {
		for idx := uint32(0); idx < (1 << ptttype.HASH_BITS); idx++ {
			checkHash(idx)
		}
	}
}

func checkHash(h uint32) {
	// p as delegate-pointer to the Shm.
	// in the beginning, p is the indicator of HashHead.
	// after 1st for-loop, p is in nextInHash.
	// val as the corresponding *p

	// line: 71
	p := h
	val := int32(0)
	pval := &val
	valptr := unsafe.Pointer(pval)
	Shm.ReadAt(
		unsafe.Offsetof(Shm.HashHead)+types.INT32_SZ*uintptr(p),
		types.INT32_SZ,
		valptr,
	)
	if val != -1 {
		log.Infof("checkHash: after read at HashHead: h: %v p: %v val: %v", h, p, val)
	}

	// line: 72
	isFirst := true

	var offset uintptr = 0
	offsetHashHead := unsafe.Offsetof(Shm.HashHead)
	offsetNextInHash := unsafe.Offsetof(Shm.NextInHash)

	userID := [ptttype.IDLEN + 1]byte{}
	deep := 0
	for val != -1 {
		offset = offsetNextInHash
		if isFirst {
			offset = offsetHashHead
		}

		// check invalid pointer-val, set as -1  line: 74
		if val < -1 || val >= ptttype.MAX_USERS {
			log.Warnf("uhash_loader.checkHash: val invalid: h: %v p: %v val: %v isHead: %v", h, p, val, isFirst)
			*pval = -1
			Shm.WriteAt(
				offset+types.INT32_SZ*uintptr(p),
				types.INT32_SZ,
				valptr,
			)
			break
		}

		// get user-id: line: 75
		Shm.ReadAt(
			unsafe.Offsetof(Shm.Userid)+ptttype.USER_ID_SZ*uintptr(val),
			ptttype.USER_ID_SZ,
			unsafe.Pointer(&userID),
		)
		log.Infof("checkHash: (in-for-loop): after read userID: h: %v p: %v val: %v userID: %v", h, p, val, types.CstrToString(userID[:]))

		userIDHash := cmsys.StringHash(userID[:]) % (1 << ptttype.HASH_BITS)

		// check hash as expected line: 76
		if userIDHash != h {
			// XXX
			// the result of the userID does not fit the h (broken?).
			// XXX uhash_loader is used only 1-time when starting the service.
			next := int32(0)

			// get next from *p (val)
			Shm.ReadAt(
				offsetNextInHash+types.INT32_SZ*uintptr(val),
				types.INT32_SZ,
				unsafe.Pointer(&next),
			)
			log.Warnf("userID hash is not in the corresponding idx (to remove) (%v): userID: %v userIDHash: %v h: %v next: %v", deep, types.CstrToString(userID[:]), userIDHash, h, next)
			// remove current by setting current as the next, hopefully the next user can fit the userIDHash.
			*pval = next
			Shm.WriteAt(
				offset+types.INT32_SZ*uintptr(p),
				types.INT32_SZ,
				unsafe.Pointer(&next),
			)
		} else {
			// 1. p as val (pointer in NextInHash)
			// 2. update val as NextInHash[p]
			p = uint32(val)
			Shm.ReadAt(
				offsetNextInHash+types.INT32_SZ*uintptr(p),
				types.INT32_SZ,
				unsafe.Pointer(&val),
			)
			isFirst = false

			log.Infof("checkHash: (in-for-loop (match)): after read next: h: %v p: %v val: %v userID: %v isFirst: %v", h, p, val, types.CstrToString(userID[:]), isFirst)
		}

		// line: 87
		deep++
		if deep == 100 {
			log.Warnf("checkHash deadlock: deep: %v h: %v p: %v val: %v isFirst: %v", deep, h, p, val, isFirst)
			break
		}
	}
}
